### PR TITLE
feat(typescript): Align `isolatedDeclaration` implementation with tsc

### PR DESCRIPTION
**Pull request by CPunisher**
**Description:**
- [x] **Basic implementation.** This also includes diagnostics and type inference.
    - [x] Function
    - [x] Class
    - [x] Variable declaration
    - [x] Enum
- [x] **Remove unused imports and declarations.** Based on the result ast of fast dts transformation, we collect the usage of ids which also contain syntax context. Then we prune the ast to cut off the unused imports and declarations.
- [x] **Port tests and bug fixes.** To ensure correctness, I'd like to port the fixtures as oxc does https://github.com/oxc-project/oxc/tree/main/crates/oxc_isolated_declarations/tests/fixtures.
- [x] **Benmarks**
- [x] Strip internal

**Implementation:** 
Most code is direct translation of https://github.com/oxc-project/oxc/tree/main/crates/oxc_isolated_declarations. The differences come from:
1. swc and oxc use **different ast**.
2. The transformation of swc is **mutation-based**, which directly mutates the whole cloned ast, while the transformation of oxc is **immutation-based**, which constructs and copies the ast nodes on demand. Maybe mutation-based transformation is not better, but I think it's also annoyed to construct ast nodes and it could be easy to refactor if there is some demands in the future.
3. oxc transforms while collecting references information to prune unused idents. However, the ast nodes of swc contain **syntax context**, we don't need to collect the information again. So I build a reference graph and calculate the reachability of identifiers.
4. I think oxc could have bugs. I manually compare the result with `tsc --declaration --isolatedDeclarations --emitDeclarationOnly --isolatedModules --noResolve --noCheck --strictNullChecks test.ts`. I also reguard those cases where tsc can't compile as undefined behaviors.

😅 Actually I find my implementation of point 2 is not too good either. I meet many bad cases after my basic implementation and write many ugly patches. So issues are welcome.

**Benchmarks:**
Roughly test with https://raw.githubusercontent.com/oxc-project/benchmark-files/main/vue-id.ts under M3Pro.
|  | parse without syntax context resolution | isolated declarations |
| -------------- | ------ | --------------------------------- |
| swc | 282.68ms | 335.44ms |
| oxc | 53.286ms | 63.917ms | 
| tsc | - | 3.875s |

**Known problem:**
- Due to the comment emitting algorithm in swc, isolated declaration may also generate comments with wrong positions.

**Related issue (if exists):**
closes https://github.com/swc-project/swc/issues/9705
closes https://github.com/swc-project/swc/issues/9718

### CODE REVIEW

**Comment by SoonIter** [comment]
thanks for your help❤️️
 
😂 got an another blocking point of isolatedDeclaration

should generate the same error as tsc

```
src/App.tsx(8,10): error TS9007: Function must have an explicit return type annotation with --isolatedDeclarations.
```

I'll open another issue

**Comment by Boshen** [comment]
FYI oxc uses immutable AST for a single transform pipeline.

**Comment by CPunisher** [comment]
> FYI oxc uses immutable AST for a single transform pipeline.

I see. And I believe immutable AST gets better performance and other engineering benefits in isolated declaration transformation. After finishing, I have a little bit of regret... But maybe in the next pr 😁.

File: crates/swc/Cargo.toml [hunk 1]
```diff
@@ -48,8 +48,8 @@ plugin_transform_host_native = [
   "swc_plugin_runner/plugin_transform_host_native",
 ]
 # Do not inject tokio runtime while running plugin transforms
-manual-tokio-runtmie = ["manual-tokio-runtime"]
 manual-tokio-runtime = []
+manual-tokio-runtmie = ["manual-tokio-runtime"]
```
Comment (kdy1) [inline crates/swc/Cargo.toml:52]: Can you revert this?

File: crates/swc/benches/isolated_declarations.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    swc::Compiler::new(cm)
+}
+
+fn bench_isolated_declarations(criterion: &mut Criterion) {
+    criterion.bench_function("isolated-declarations", |b| {
+        let c = mk();
+        let fm =
+            c.cm.load_file(Path::new("benches/assets/vue-id.ts"))
+                .unwrap();
+        b.iter_with_large_drop(|| {
```
Comment (kdy1) [inline crates/swc/benches/isolated_declarations.rs:20]: `iter_with_large_drop` is effective only if you return the variable with large drop.

Comment (CPunisher) [reply crates/swc/benches/isolated_declarations.rs]: Yes, I find there is no difference if I change it to `iter`.

**Comment by kdy1** [comment]
Sorry for the delay. I'm still reviewing this.

**Review by kdy1 — APPROVED** [review]

**Review by kdy1 — APPROVED** [review]
Thank you so much for a great work!

**Pull request merged**